src/Raven.Client/Documents/LinqExtensions.cs

```diff
@@ -313,6 +313,19 @@ public static Task<List<T>> ToListAsync<T>(this IQueryable<T> source, Cancellati
             return documentQuery.ToListAsync(token);
         }
 
+        /// <summary>
+        /// Returns an array of results for a query asynchronously. 
+        /// </summary>
+        public static Task<T[]> ToArrayAsync<T>(this IQueryable<T> source, CancellationToken token = default)
+        {
+            var provider = source.Provider as IRavenQueryProvider;
+            if (provider == null)
+                throw new ArgumentException("You can only use Raven Queryable with ToArrayAsync");
+
+            var documentQuery = provider.ToAsyncDocumentQuery<T>(source.Expression);
+            return documentQuery.ToArrayAsync(token);
```

Comment:
@valeriob It is actually a big change, because now all queries that were doing `ToList()` will be allocating an array and after that allocating a second list.

I think it would be better to just invoke `ToListAsync` extension method here and issue `ToArray` on the results here as a shortcut.

Reply:
Yes and no: everybody will allocate an array , and only who calls ToList() will have the overhead of allocating a new list from it. 
I think that query should as default return arrays and not lists (like Load() does) and ppl should opt in into List or other structures. I've never used ToList() in any db (ORM, uORM or ravendb 2/3/4) query. I think that ravendb4 i so relative new that this change is still possible, maybe we can add some comment on the ToList version to warn ppl of that allocation, even though it should be expected 😄.
Consider also that the gains obtained from initializing the array upfront should be more than the performances spent doing a single copy in case of ToList(), so i think it's a net gain afterall (i did not measure, just counting copies of memory, consider that EnsureCapacity doubles the array every time, so can be pretty inefficient).

https://referencesource.microsoft.com/#mscorlib/system/collections/generic/list.cs,eb66b6616c6fd4ef,references
https://referencesource.microsoft.com/#mscorlib/system/collections/generic/list.cs,9808f1f5ef16c436